switchy/apps/__init__.py

```diff
@@ -1,6 +1,112 @@
+# vim: tabstop=4 softtabstop=4 shiftwidth=4 textwidth=80 smarttab expandtab
 # This Source Code Form is subject to the terms of the Mozilla Public
 # License, v. 2.0. If a copy of the MPL was not distributed with this
 # file, You can obtain one at http://mozilla.org/MPL/2.0/.
 '''
 Built-in applications
 '''
+from .. import utils
+from collections import OrderedDict
+import itertools
+import operator
+
+# apps registry
+_apps = OrderedDict()
+
+
+def app(*args, **kwargs):
+    '''Decorator to register switchy application classes.
+       Example usage:
+
+       @app
+       class CoolAppController(object):
+           pass
+
+       This will register the class as a switchy app.
+       The name of the app defaults to `class.__name__`.
+       The help for the app is taken from `class.__doc__`.
+
+       You can also provide an alternative name via a
+       decorator argument:
+
+       @app('CoolName')
+       class CoolAppController(object):
+           pass
+
+       or with a keyword arg:
+
+       @app(name='CoolName')
+       class CoolAppController(object):
+           pass
+    '''
+    name = kwargs.get('name')
+    if len(args) >= 1:
+        arg0 = args[0]
+        if type(arg0) is type:
+            return register(arg0, None)
+        name = arg0
+        # if len(args) > 1:
+        #     tail = args[1:]
+
+    def inner(cls):
+        return register(cls, name=name)
+    return inner
+
+
+def register(cls, name=None):
+    name = name or cls.__name__
+    try:
```

Comment:
Hmm I guess the only slightly more clever way to do this would be:
```python
def register(cls, name=None):
    app = _apps.setdefault(name or cls.__name__, cls)
    if cls is not app:
        raise ValueError("An app '{}' already exists with name '{}'"
                                    .format(app, name))
    return app
```
What do you think? I can change it.